### PR TITLE
`single_match`: fix checking of explicitly matched enums

### LINKED ISSUE #11365: `single_match` incorrectly lints on exhaustive matches inside of Option/Result
### Summary

`single_match` isn't supposed to lint on exhaustive matches of user-defined types, but it will if those are inside an Option or Result.

### Lint Name

single_match

### Reproducer

I tried this code:

```rust
enum Test {
    A,
    B,
}

fn main() {
    let test = Some(Test::A);

    match test {
        Some(Test::A | Test::B) => todo!(),
        None => {}
    }
}

```

Clippy suggests this code:

```rust
    if let Some(Test::A | Test::B) = test {
        todo!()
    }

```

I expected to see this happen:

Nothing, `single_match` should not trigger here since I've exhaustively listed every variant of my enum. It seems some specific carve-outs were added for Option/Result in `single_match`, but they're over-eager. The code clippy suggests is equivalent, right now, but if I were to add a `Test::C` variant without remembering to handle it, clippy's version would not catch it.

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #11365

This approach has false-negatives, but fixing them will require a significant amount of additional state tracking. The comment in `add_and_pats` has the explanation.

changelog: `single_match`: correct checking if the match explicitly matches all of an enum's variants.

### CODE REVIEW

**Comment by Jarcho** [comment]
r? clippy

**Comment by xFrednet** [comment]
Hey @dswij , this is a ping from triage. Can you give this PR a review? It's totally fine if you don't have the time right now, you can reassign the PR to a random team member using `r? clippy`.

@rustbot ready

**Comment by dswij** [comment]
I won't be back until Wednesday. If anyone have the capacity, feel free to review this :)

r? clippy

**Comment by xFrednet** [comment]
Hey, this is a ping from triage. @blyxyas  can you give this PR a review? It's totally fine if you don't have the time right now, you can reassign the PR to a random team member using `r? clippy`.

@rustbot ready

**Comment by blyxyas** [comment]
I've been postponing it for some time, I'll take a deep look at this tomorrow (compromise), if then it doesn't click, I'll do a reroll

File: clippy_lints/src/matches/single_match.rs [hunk 1]
```diff
@@ -122,133 +124,231 @@ fn report_single_pattern(
...
-    // have more variants in the future.
-    if form_exhaustive_matches(cx, ty, arms[0].pat, arms[1].pat) {
-        report_single_pattern(cx, ex, arms, expr, els);
+struct PatVisitor<'tcx> {
+    typeck: &'tcx TypeckResults<'tcx>,
+    has_binding: bool,
+    has_enum: bool,
+}
+impl<'tcx> Visitor<'tcx> for PatVisitor<'tcx> {
+    fn visit_pat(&mut self, pat: &'tcx Pat<'_>) {
```
Comment (blyxyas) [inline clippy_lints/src/matches/single_match.rs:151]: You can use ControlFlow here and replace has_binding with `.is_break()`

```suggestion
    type Result = ControlFlow<()>
    fn visit_pat(&mut self, pat: &'tcx Pat<'_>) -> ControlFlow<()> {
```

File: clippy_lints/src/matches/single_match.rs [hunk 2]
```diff
@@ -122,133 +124,231 @@ fn report_single_pattern(
...
+        if self.has_binding {
+            return;
+        }
+        self.has_enum |= self.typeck.pat_ty(pat).ty_adt_def().map_or(false, AdtDef::is_enum);
+        self.has_binding = matches!(pat.kind, PatKind::Binding(..));
```
Comment (blyxyas) [inline clippy_lints/src/matches/single_match.rs:156]:
```suggestion
        if matches!(pat.kind, PatKind::Binding(..)) { return ControlFlow::Break(()) } else { walk_pat(self, pat) };
```

File: clippy_lints/src/matches/single_match.rs [hunk 3]
```diff
@@ -122,133 +124,231 @@ fn report_single_pattern(
...
+    typeck: &'tcx TypeckResults<'tcx>,
+    has_binding: bool,
+    has_enum: bool,
+}
+impl<'tcx> Visitor<'tcx> for PatVisitor<'tcx> {
+    fn visit_pat(&mut self, pat: &'tcx Pat<'_>) {
+        if self.has_binding {
+            return;
+        }
+        self.has_enum |= self.typeck.pat_ty(pat).ty_adt_def().map_or(false, AdtDef::is_enum);
```
Comment (blyxyas) [inline clippy_lints/src/matches/single_match.rs:155]:
```suggestion
        self.has_enum |= self.typeck.pat_ty(pat).ty_adt_def().map_or(false, AdtDef::is_enum);
```

File: clippy_lints/src/matches/single_match.rs [hunk 4]
```diff
@@ -27,68 +29,68 @@ fn empty_arm_has_comment(cx: &LateContext<'_>, span: Span) -> bool {
...
+                typeck,
+                has_binding: false,
+                has_enum: false,
+            };
+            v.visit_pat(arm2.pat);
+            if v.has_binding {
+                return;
+            }
+            if v.has_enum {
+                let cx = PatCtxt {
```
Comment (blyxyas) [inline clippy_lints/src/matches/single_match.rs:67]: (See suggestion about using ControlFlow below)
```suggestion
            if !v.visit_pat(arm2.pat).is_break() && v.has_enum {
                let cx = PatCtxt {
```

File: clippy_lints/src/matches/single_match.rs [hunk 5]
```diff
@@ -122,133 +124,231 @@ fn report_single_pattern(
...
-            true
-        },
-        (PatKind::TupleStruct(..), PatKind::Path(_)) => pat_in_candidate_enum(cx, ty, right),
-        (PatKind::TupleStruct(..), PatKind::TupleStruct(_, inner, _)) => {
-            pat_in_candidate_enum(cx, ty, right) && inner.iter().all(contains_only_wilds)
-        },
-        _ => false,
+            },
+        }
     }
```
Comment (blyxyas) [inline clippy_lints/src/matches/single_match.rs:353]: Could you add some comments here? I'm having trouble understanding the meaning behind all these branches (and the meaning behind the function in general)

**Comment by xFrednet** [comment]
Hey, this is triage: It looks like @blyxyas is currently busy, let's pick a new reviewer.

r? clippy

**Comment by Manishearth** [comment]
r? clippy

(in mumbai for two weeks, not using the computer as much)

File: clippy_lints/src/matches/single_match.rs [hunk 6]
```diff
@@ -114,124 +123,245 @@ fn report_single_pattern(
...
+    ) -> Option<(&mut Self, &'tcx VariantDef)> {
+        let states = match self {
+            Self::Wild => return None,
+            Self::Other => {
+                *self = Self::StdEnum(cx.arena.alloc_from_iter((0..adt.variants().len()).map(|_| Self::Other)));
+                let Self::StdEnum(x) = self else {
+                    unreachable!();
+                };
+                x
+            },
```
Comment (xFrednet) [inline clippy_lints/src/matches/single_match.rs:232]: This confuses me. The description of `Other` states:

> For non-std enums [...]

Is this intentional? If so, could you add a comment explaining why it's safe to assume that `Other` is an `StdEnum`.

**Review by xFrednet — COMMENTED** [review]
One small comment, the rest looks good to me.

File: clippy_lints/src/matches/single_match.rs (hunk 6 above)
Comment (Jarcho) [reply clippy_lints/src/matches/single_match.rs:232]: From the first line describing other: "Either the initial state for a pattern or a non-std enum".

`get_enum_variant` is called to get the state of a variant in a std enum.

Comment (Jarcho) [reply clippy_lints/src/matches/single_match.rs:234]: The function name has been changed to `get_std_enum_variant` to clarify.

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
LGTM, thanks!

---

Roses are red,
Violets are blue,
This PR has been open,
and should be merged now

**Pull request merged**